R/integrate_sces.R

```diff
@@ -0,0 +1,180 @@
+                                        batch_column,
+                                        covariate_cols,
+                                        ...)
+  }
+
+  # Add PCs from integration into merged_sce (name is lowercase)
+  reducedDim(merged_sce, glue::glue("{integration_method}_PCA")) <- integrated_pcs
+
+  # Return the merged_sce with integration results included
+  return(merged_sce)
+}
+
+
+
+#' Integrate a list of SingleCellExperiment (SCE) objects using the fastMNN
+#' approach from the `batchelor` package.
+#' Source: http://www.bioconductor.org/packages/release/bioc/html/batchelor.html
+#'
+#' @param merged_sce A merged SCE object as prepared by `scpcaTools::merge_sce_list()`.
+#' @param batch_column The variable in `merged_sce` indicating batches.
+#' @param ... Additional arguments to pass into `batchelor::fastMNN()`
+#'
+#' @return An integrated SCE object produced by `fastMNN`
+#'
+#' @import SingleCellExperiment
+integrate_fastmnn <- function(merged_sce,
+                              batch_column,
+                              ...) {
+
+  # Check the merged_sce for logcounts
+  if (!("logcounts" %in% names(assays(merged_sce)))) {
+    stop("The `merged_sce` object requires a `logcounts` assay for fastMNN integration.")
+  }
+
+  # Perform integration
+  integrated_sce <- batchelor::fastMNN(merged_sce,
+                                       batch = colData(merged_sce)[,batch_column],
+                                       ...)
+
+  return(integrated_sce)
+}
+
+
+
+
+#' Integrate a list of SingleCellExperiment (SCE) objects using the harmony
+#' approach from the `harmony` package.
+#' Source: https://cran.r-project.org/web/packages/harmony/index.html
+#'
+#' @param merged_sce A merged SCE object as prepared by `scpcaTools::merge_sce_list()`.
+#' @param batch_column The column in the merged SCE object indicating batches
+#'   being integrated.
+#' @param covariate_cols A vector of other columns to consider as
+#'   covariates during integration.
+#' @param ... Additional arguments to pass into `harmony::harmonyMatrix()`
+#'
+#' @return Integrated PCs as calculated by `harmony`
+#'
+#' @import SingleCellExperiment
+integrate_harmony <- function(merged_sce,
+                              batch_column,
+                              covariate_cols = c(),
+                              ...) {
+
+  # Check the merged_sce
+  if (!("PCA" %in% reducedDimNames(merged_sce))) {
+    stop("The merged_sce object requires a `PCA` reduced dimension for harmony integration.")
+  }
+
+  # Check covariate columns
+  if (!(all(covariate_cols %in% names(colData(merged_sce))))) {
+    stop("The provided covariate columns are not all present in the `merged_sce` colData.")
+  }
+
+  # Setup harmony metadata
+  covariate_cols <- c(batch_column, covariate_cols)
+  harmony_metadata <- tibble::as_tibble(colData(merged_sce)) %>%
+    dplyr::select(
+      dplyr::all_of(c(batch_column, covariate_cols))
```

Comment:
I think this is redundant since batch_column was already added?
```suggestion
      dplyr::all_of(covariate_cols)
```